design/open/kv3-execution-plan.md

```diff
@@ -0,0 +1,52 @@
+
+# General
+## Tables
+Graveler uses 5 DB tables:
+* `graveler_repositories`,`graveler_branches`,`graveler_commits`,`graveler_tags` - used by `pkg/graveler/ref`
+* `graveler_staging_kv` - used by `pkg/graveler/staging`
+
+`graveler_repositories`, `graveler_commits` and `graveler_tags` are immutable and does not update entries, and so should be straight forward to implement over our `KV` (`graveler_commits` actually performs update for commit generation, but this is only valid for `refs-restore` which can only take place on a `bare` repo and so it is of no concern)</br>
+`graveler_branches` is currently being protected by a lock, during commit actions, and in order to support lockless commits with `KV`, the [following design](https://github.com/treeverse/lakeFS/blob/b3204edad00f88f8eb98524ad940fde96e02ab0a/design/open/metadata_kv/index.md#graveler-metadata---branches-and-staged-writes) is proposed. However, as a first step we can implement it using a naive table lock, just to get it going, or better, a `branch-key` based locking, to prevent staging during commits</br>
```

Comment:
After reading it once again with the help of @arielshaqed and @N-o-Z comments, I think I have a better idea of what you were aiming for here. A `naive table lock` means that we would be using the same locking technique we have for the current postgres DB, right? Well - that cannot be achieved with KV. It's not part of the KV interface and adding it to the current 3 implementations we have is harder than just straightforward use the mechanisms described in the original proposal.

Reply:
I think by now we are arguing about different words to describe the exact same thing.  I would strongly encourage that we simply stop using the word "lock", CAS-type operations ("SetIf") like those used here are at a lower level and easier to use directly.  The term "lock" is overloaded and has many different meaning.  (Note that "lock-free" does **not** talk about locks, intentionally, so we can continue to use that term.)

Let's be concrete.  I call the following "the na&iuml;ve way" to perform a protected action on a branch:

1. Get the original branch state `o`.
1. (Maybe perform some actions, they might all be lost in the last step.)
1. Compute a new branch state `n`.
1. SetIf the branch state to `n` IF it is still `o`.
   * (If setting failed and some other actions were performed, maybe clean them up.)

This type of process is sometimes loosely called "optimistic locking".

Reply:
Actually, what I had in mind for **naive and temporary locking** is locking of some common resource, on a logical level, not related to `KV`.
Both staging and commit operations (on the same branch) mutually block each other (this is the case, right? any other operatios need to be blocked), by locking a common resource (mutex?), just for the correctness of things.
We can reduce the contention by having NumLocks (say 2^10) such locks, and each operatation locks branchID%NumLocks, letting several operations (up to NumLocks, not sure how to calculate the average) run in parallel.
Any reason why that should not work as a **naive, temporary yet correct, solution**?